src/java/main/org/apache/zookeeper/server/NettyServerCnxn.java

```diff
@@ -100,11 +101,15 @@ public void close() {
                         + Long.toHexString(sessionId));
             }
 
-            synchronized (factory.ipMap) {
-                Set<NettyServerCnxn> s =
-                    factory.ipMap.get(((InetSocketAddress)channel
-                            .getRemoteAddress()).getAddress());
-                s.remove(this);
+            InetAddress address = ((InetSocketAddress) channel.getRemoteAddress()).getAddress();
+            Set<NettyServerCnxn> s = factory.ipMap.get(address);
```

Comment:
I agree that that ipMap management would be cleaner if it was handled in the factory.